src/components/Auth/CP/CPAuthProvider.tsx

```diff
@@ -0,0 +1,70 @@
+import CPAuth from 'lib/CPAuth/CPAuth';
+import { OAuth2Events } from 'lib/OAuth2/OAuth2';
+import { IOAuth2User } from 'lib/OAuth2/OAuth2User';
+import React, { ReactElement, useCallback, useEffect } from 'react';
+import { useDispatch } from 'react-redux';
+import {
+  loadUserSuccess,
+  userExpired,
+  userExpiring,
+  userSignedOut,
+} from 'stores/cpauth/actions';
+
+interface ICPAuthProviderProps extends React.PropsWithChildren<{}> {}
+
+const CPAuthProvider: React.FC<ICPAuthProviderProps> = ({ children }) => {
+  const dispatch = useDispatch();
+
+  const onUserLoaded = useCallback(
+    (event: CustomEvent<IOAuth2User | null>) => {
+      dispatch(loadUserSuccess(event.detail));
+    },
+    [dispatch]
+  );
+
+  const onUserSignedOut = useCallback(() => {
+    dispatch(userSignedOut());
+  }, [dispatch]);
+
+  const onTokenExpiring = useCallback(() => {
+    dispatch(userExpiring());
+  }, [dispatch]);
+
+  const onTokenExpired = useCallback(() => {
+    dispatch(userExpired());
+  }, [dispatch]);
+
+  useEffect(() => {
+    const cpAuth = CPAuth.getInstance();
+    cpAuth.addEventListener(OAuth2Events.UserLoaded, onUserLoaded);
+    cpAuth.addEventListener(OAuth2Events.UserUnloaded, onUserSignedOut);
+    cpAuth.addEventListener(OAuth2Events.UserSignedOut, onUserSignedOut);
+    cpAuth.addEventListener(OAuth2Events.TokenExpiring, onTokenExpiring);
+    cpAuth.addEventListener(OAuth2Events.TokenExpired, onTokenExpired);
+    cpAuth.addEventListener(OAuth2Events.SilentRenewError, onUserSignedOut);
+
+    return () => {
+      onUserSignedOut();
```

Comment:
When this component is unmounted, it means that we navigated to a page outside the logged in context, which means that we can safely log out.